### ISSUE TITLE
no-eval don't match on `window.eval()`

### ISSUE THREAD

**Issue by BaggersIO** [eslint/eslint#4399]
Hey,

the `no-eval` rule works fine with `eval()`, but eslint don't detect `window.eval()` expressions.

My rule:
```
"no-eval" 2
```

My env:
```
"browser": true
```

ESLint version: `1.8.0`

Cheers,
Peter

**Comment by eslintbot** [comment]
Thanks for the issue! If you're reporting a bug, please be sure to include:

1. The version of ESLint you are using (run `eslint -v`)
2. What you did (the source code and ESLint configuration)
3. The actual ESLint output complete with numbers
4. What you expected to happen instead

Requesting a new rule? Please see [Proposing a New Rule](http://eslint.org/docs/developer-guide/contributing/new-rules) for instructions.

**Comment by michaelficarra** [comment]
It looks like no-eval only detects direct calls to eval, which makes sense, since those are the only ones that give you dynamic scope and kill performance.

**Comment by platinumazure** [comment]
@michaelficarra Wait, `window.eval` does not evaluate in global/dynamic scope? And I can't imagine the `eval` variable lookup is the rate-limiting step when calling that function.

**Comment by michaelficarra** [comment]
@platinumazure You should look into the difference between direct/indirect eval.

Direct eval:

```js
(function() {
  var x = 0;
  eval("x"); // 0
})();
```

```js
(function() {
  eval("var y = 0");
  y; // 0
})();
```

Indirect eval:

```js
(function() {
  var x = 0;
  (0, eval)("x"); // ReferenceError
})();
```

```js
(function() {
  (0, eval)("var x = 0");
  x; // ReferenceError
})();
```

Performance dies because you pretty much can't predict anything at all about what's going to happen when direct eval is used, so the JS engine will drop down into a naïve interpreter mode.

**Comment by platinumazure** [comment]
Crazy. Learn something new every day. Does `window.eval` count as an indirect eval?

**Comment by michaelficarra** [comment]
Yes. A direct call to eval must appear as a call to the identifier `eval`, and it must resolve to the original value of the `eval` property of the global object (called `%eval%` in the spec).

```js
(function(eval) {
  eval(...) // direct call to eval
})(eval);
```

```js
(function(eval) {
  eval(...) // indirect call to eval
})(function(x){
  return eval(x); // direct call to eval
});
```

**Comment by nzakas** [comment]
This rule is about blocking all uses of `eval`, so it makes sense to block `window.eval` as well.

**Comment by michaelficarra** [comment]
Then an option should be added to allow safe forms of eval.

**Comment by nzakas** [comment]
Can you file an issue for that?

**Comment by michaelficarra** [comment]
Also, it is impossible to detect all indirect calls to eval statically. Where are you going to draw the line?

**Comment by michaelficarra** [comment]
"If we ever change no-eval to detect indirect calls to eval, we should add an option to disable it"? I'm not creating an issue for that, it should be discussed here.

**Comment by nzakas** [comment]
This issue is a bug, we did not mean to miss window.eval. we will fix that bug and this issue will be closed. You're requesting a new option that distinguishes between direct and indirect eval, which is an enhancement, so we need a separate issue.

**Comment by mysticatea** [comment]
I will work on this.

My plan:

- Checks `evel` property of the host object.
  - When `env: "browser"` is specified, the host object is `window`.
  - When `env: "node"` is specified, the host object is `global`.
  - Other idea?
- Checks indirect-calls in an expression.
  - `(0, eval)(...)`
  - `[eval][0](...)`
  - `{a: eval}.a(...)`
- Does NOT check indirect-calls via other variables. To detect those is too hard.
  - `var foo = eval; foo(...)`
  - `(function(foo) { foo(...) }(eval))`
  - `var i = 0; [eval][i](...)`
  - `var eval = "eval"; window[eval](....)`

**Comment by michaelficarra** [comment]
How about top-level `this.eval` in scripts? `eval.call(null, "...")`? `Function.prototype.call.call(eval, null, "...")`?

**Comment by michaelficarra** [comment]
@nzakas The rule currently distinguishes between direct/indirect eval, whether intentional or not. If you remove this capability in a bugfix, those that are okay with indirect eval and not direct eval (and it is not unreasonable to have this preference) will have no way to configure the rule for their preference. I think the best way to move forward with this *for eslint users* is to give an option to preserve the current functionality in the same release it is removed. I can open a separate issue if you still want, but I strongly believe it should be a blocker for this issue.

**Comment by mysticatea** [comment]
Oh, I had not noticed those. `eval.bind(null, "...")()`, haha, I have a headache.

I'm going to change strategy.

- `no-eval` just checks the variable object of the global `eval`.
  Then reports all references (contains readonly) of the variable.
  This is blocking most indirect `eval` calls.
- Checks `evel` property of the host object.
  - When `env: "browser"` is specified, the host object is `window`.
  - When `env: "node"` is specified, the host object is `global`.
  - `no-eval` reports *accesses*, not calls.
- Checks `this.eval` for functions (not methods, not strict mode).
  Hmm, I want to share function's info with `no-invalid-this` rule.
  Or may I not check `this.eval`, because `no-invalid-this` is catching those?

When allowing indirect calls (#4441), the detection logic is [as is](https://github.com/eslint/eslint/blob/master/lib/rules/no-eval.js).
In short, all `eval()` calls are reported.

And I will write this known limitation in the doc.

**Comment by mysticatea** [comment]
No, `no-invalid-this` does not check sloppy mode code....

**Comment by nzakas** [comment]
@michaelficarra every bug fix is backwards incompatible, we don't always need to keep previous behavior if it was incorrect.

**Comment by nzakas** [comment]
Oops, hit wrong button.

**Comment by michaelficarra** [comment]
@nzakas Sorry, that only works when the observed ("incorrect") behaviour could not have been reasonably expected as intentional behaviour. Since the current behaviour could be reasonably expected (if you thought "eval" meant "direct eval") and *useful*, it should be preserved. Anyway, this shouldn't matter as it seems @mysticatea is already handling it this way.

**Comment by mysticatea** [comment]
Oh...
I cannot get `env` setting from `RuleContext`...

**Comment by nzakas** [comment]
@michaelficarra you're assuming users have a much deeper understanding of the language than is typical. When we say "disallow eval", I'd be shocked to see any number of people believe that what we really meant was "disallow direct eval." I've already agreed to add an option so you can make such a distinction if you want, so I'm not sure why you're still debating this.

**Comment by ilyavolodin** [comment]
I agree with @nzakas While @michaelficarra has a good point, I know for a fact that every JavaScript developer I ever worked with does not know the distinction between direct and indirect eval, it's just not something you run into on a daily bases.
